sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1013 @@
+        if verbose:
+            print('Computing static part of the gradient...')
+
+        n_samples, n_neighbors = targets.shape
+        row = np.repeat(range(n_samples), n_neighbors)
+        col = targets.ravel()
+        targets_sparse = csr_matrix((np.ones(targets.size), (row, col)),
+                                    shape=(n_samples, n_samples))
+
+        return sum_outer_products(X, targets_sparse)
+
+    def _lbfgs_callback(self, transformation):
+        if self.callback is not None:
+            self.callback(transformation, self.n_iter_)
+
+        self.n_iter_ += 1
+
+    def _loss_grad(self, transformation, X, y, targets, grad_static):
+        """Compute the loss under a given ``transformation`` and the
+        loss gradient w.r.t. ``transformation``.
+
+        Parameters
+        ----------
+        transformation : array, shape (n_features_out * n_features,)
+            The current (flattened) linear transformation.
+
+        X : array-like, shape (n_samples, n_features)
+            The training samples.
+
+        y : array, shape (n_samples,)
+            The corresponding training labels.
+
+        targets : array, shape (n_samples, n_neighbors)
+            The target neighbors of each sample.
+
+        grad_static : array, shape (n_features, n_features)
+            The gradient component caused by target neighbors, that stays
+            fixed throughout the algorithm.
+
+        Returns
+        -------
+        loss: float
+            The new loss.
+        grad: array, shape (n_features_out * n_features,)
+            The new (flattened) gradient.
+        """
+
+        n_samples, n_features = X.shape
+        self.transformation_ = transformation.reshape(-1, n_features)
+
+        t_start = time.time()
+        X_embedded = self.transform(X, check_input=False)
+
+        # Compute squared distances to target neighbors (plus margin)
+        n_neighbors = targets.shape[1]
+        dist_tn = np.zeros((n_samples, n_neighbors))
+        for k in range(n_neighbors):
+            dist_tn[:, k] = row_norms(X_embedded - X_embedded[targets[:, k]],
+                                      squared=True) + 1
+
+        # Compute impostors and squared distances to them
+        margin_radii = dist_tn[:, -1] + 1
+        imp_row, imp_col, dist_imp = \
+            self._find_impostors(X_embedded, y, margin_radii, self.use_sparse)
+
+        loss = 0
+        shape = (n_samples, n_samples)
+        A0 = csr_matrix(shape)
+        for k in reversed(range(n_neighbors)):
+            loss1 = np.maximum(dist_tn[imp_row, k] - dist_imp, 0)
+            ac, = np.where(loss1 > 0)
+            A1 = csr_matrix((2*loss1[ac], (imp_row[ac], imp_col[ac])), shape)
+
+            loss2 = np.maximum(dist_tn[imp_col, k] - dist_imp, 0)
+            ac, = np.where(loss2 > 0)
+            A2 = csc_matrix((2*loss2[ac], (imp_row[ac], imp_col[ac])), shape)
+
+            values = np.squeeze(np.asarray(A1.sum(1).ravel() + A2.sum(0)))
```

Comment:
use prefer to avoid squeeze and collapse axis explicitly

Reply:
I can use `(A1.sum(1).ravel() + A2.sum(0)).getA1()` although `getA1()` is not that descriptive.

Reply:
I prefer this. Again squeeze is dangerous. It can collapse more than one dimension and it's better to avoid it when possible.